power_grid_model_c/power_grid_model/include/power_grid_model/main_core/input.hpp

```diff
@@ -158,6 +163,13 @@ inline void add_component(MainModelState<ComponentContainer>& state, ForwardIter
             emplace_component<Component>(state, id, input, regulated_object_type, u_rated);
         }
     }
+    // Make sure that each regulated object has at most one regulator
+    const std::unordered_set<Idx2D, Idx2DHash> unique_regulated_objects(regulated_objects.begin(),
+                                                                        regulated_objects.end());
+    if (unique_regulated_objects.size() != regulated_objects.size()) {
+        // There are duplicates
+        throw DuplicatedRegulatedObject();
+    }
```

Comment:
is this faster than just looping over the data?

Reply:
'just looping over the data' reads quadratic complexity ;)

Reply:
yep i realized after i posted this 😬

Reply:
maybe you can create a utility for this? like `has_duplicates(Range, Func)`

Reply:
Possibility, yes, maybe later